### PR TITLE
fix(sagemaker): parse each SSE event within a PayloadPart

### LINKED ISSUE #4035: [BUG] SageMaker streaming silently drops all content when a PayloadPart contains more than one SSE event
### Checks

- [x] I have updated to the lastest minor and patch version of Strands
- [x] I have checked the documentation and this is not expected behavior
- [x] I have searched ./issues and there are no duplicates of my issue

### SDK Language

Python

### Strands Version

`main` @ 45bcea8 (also present in python/v1.54.0)

### Language Runtime Version

Python 3.12.3

### Operating System

Linux (Ubuntu 24.04)

### Installation Method

git clone

### Steps to Reproduce

`SageMakerAIModel.stream` parses the response stream by stripping a leading `data: ` from an accumulator and calling `json.loads` on the whole buffer, resetting only when the parse succeeds — `strands-py/src/strands/models/sagemaker.py`, the `for event in response["Body"]` loop (lines 347-352 on `main`):

```python
for event in response["Body"]:
    chunk = event["PayloadPart"]["Bytes"].decode("utf-8")
    partial_content += chunk[6:] if chunk.startswith("data: ") else chunk  # TGI fix
    logger.info("chunk=<%s>", partial_content)
    try:
        content = json.loads(partial_content)
        partial_content = ""
        ...
    except json.JSONDecodeError:
        # Continue accumulating content until we have valid JSON
        continue
```

This assumes one whole JSON object per `PayloadPart`. SageMaker does not align `PayloadPart` chunks to SSE event boundaries, so as soon as the container flushes two events into one part, the accumulator becomes `{...event1...}\n\ndata: {...event2...}`, `json.loads` raises `JSONDecodeError: Extra data`, the `except` branch keeps appending, and the buffer never parses again.

To reproduce, feed the parser two complete SSE events in one `PayloadPart`, then a third event split across two parts:

```python
# PayloadPart 1: two complete "data:" events in one part
'data: {"choices":[{"delta":{"content":"Paris is "},"finish_reason":null}]}\n\n'
'data: {"choices":[{"delta":{"content":"the capital of France"},"finish_reason":null}]}\n\n'
# PayloadParts 2 and 3: a third event cut in the middle of its JSON
'data: {"choices":[{"delta":{"conten'
't":"!"},"finish_reason":"stop"}]}\n\n'
```

I drove this through a real `SageMakerAIModel` instance with a stubbed client (no AWS calls) and collected the emitted `contentBlockDelta` text.

### Expected Behavior

The stream yields `"Paris is the capital of France!"` with stop reason `end_turn`.

### Actual Behavior

The streamed text is empty. Every event is discarded, including the first one that was already complete and parseable. No exception is raised and no warning is logged — the agent just ends the turn with an empty assistant message.

The accumulator visibly grows and never resets. From the parser's own `logger.info("chunk=<%s>", partial_content)`:

```
chunk=<{"choices": [{"delta": {"content": "Paris is "}, "finish_reason": null}]}

data: {"choices": [{"delta": {"content": "the capital of France"}, "finish_reason": null}]}

>
chunk=<{"choices": [{"delta": {"content": "Paris is "}, "finish_reason": null}]}

data: {"choices": [{"delta": {"content": "the capital of France"}, "finish_reason": null}]}

{"choices": [{"delta": {"conten>
```

### Additional Context

Multi-event and partial-event framing is exactly what OpenAI-compatible serving stacks such as vLLM and TGI emit, and a SageMaker container running one of those will flush several `data:` events per part under load — so this hits real deployments rather than being a theoretical edge case.

It does not reproduce against a container that happens to send exactly one bare JSON object per `PayloadPart`, which is the shape the existing unit tests simulate. That is why the tests stay green while real streaming loses content.

The silence is the worst part: there is no error to correlate, so it looks like the model returned nothing.

### Possible Solution

Parse the stream incrementally: append bytes to a buffer, split on newlines, decode each complete `data:` line as its own event, and hold any trailing partial line until the next part completes it. Payloads that are not newline-framed (bare JSON) can still be decoded once the buffer parses as a whole, which keeps the currently working shapes behaving exactly as before.

I have a branch with that change and a regression test that fails on `main` and passes with the fix, and I would like to open a PR if the approach looks right.

### PR DESCRIPTION
**Pull request by pacocartones**
## Description

`SageMakerAIModel.stream` accumulated the response stream into one buffer, stripped a leading `data: `, and
called `json.loads` on the whole thing, only resetting the buffer when a parse succeeded. That works when each
`PayloadPart` is exactly one JSON object, but SageMaker doesn't align parts to SSE event boundaries. As soon as
a part carries two events — or one event plus the start of the next — the buffer looks like `{...}\n\ndata: {...`,
`json.loads` raises `Extra data`, and the `except` just keeps appending. The buffer never parses again, so every
remaining event is dropped in silence, including the first complete one. The stream ends with an empty assistant
message and no error.

This is easy to trip in production: multi-event and partial-event framing is exactly what OpenAI-compatible
servers like vLLM and TGI emit, and a SageMaker container running one of those will flush several `data:` events
per part under load. It doesn't reproduce against a container that happens to send one bare JSON object per part,
which is what the existing unit tests simulated — so the tests stayed green while real streaming lost content.

The fix parses the stream incrementally: append bytes to a buffer, split on newlines, decode each complete
`data:` line as its own event, and hold any trailing partial line until the next part completes it. Payloads that
aren't newline-framed (bare JSON) are still decoded once the buffer parses as a whole, so the previously working
shapes behave exactly as before.

Most of the diff is the streaming loop de-indenting by one level now that the per-event `try/except` is gone. The
new parsing lives in two small helpers, `_strip_sse_framing` and `_parse_event_stream`, which keeps `stream()`
flatter than it was.

## Related Issues

Closes #4035

## Documentation PR

Not applicable — internal parsing fix, no public API or documented behavior change.

## Type of Change

Bug fix

## Testing

Added `test_stream_multiple_events_per_payload_part` in `tests/strands/models/test_sagemaker.py`, which packs two
SSE events into one `PayloadPart` and splits a third across parts. It fails on `main` (streamed text comes back
empty) and passes with this change. The existing streaming tests, including the single-object and
split-single-object cases, still pass unchanged — 26 passed in `tests/strands/models/test_sagemaker.py`.

I also exercised the parser end to end against a `SageMakerAIModel` instance with a stubbed client and synthetic
`PayloadPart`s (no AWS calls), to confirm the multi-event and mid-JSON-split cases recover the full text and the
correct stop reason.

On `main` the parser's own log line shows the failure directly — the accumulator grows across parts and never
resets:

```
chunk=<{"choices": [{"delta": {"content": "Paris is "}, "finish_reason": null}]}

data: {"choices": [{"delta": {"content": "the capital of France"}, "finish_reason": null}]}

>
```

- [ ] I ran `hatch run prepare`

From `strands-py/`, I ran `ruff format --check` (2 files already formatted), `ruff check` (all checks passed),
`mypy src/strands/models/sagemaker.py` (no issues), and `pytest tests/strands/models/test_sagemaker.py`
(26 passed). I have not run the full `hatch run prepare` gate or the integ tests locally.

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have reviewed and understand every line of code in this PR, including any generated by AI tools, and I can explain why it works
- [x] My change is focused and reasonably small; I have split unrelated work into separate PRs
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [x] I have updated the documentation accordingly
- [x] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: strands-py/src/strands/models/sagemaker.py [hunk 1]
```diff
@@ -298,6 +298,66 @@ def format_request(
...
+    @staticmethod
+    def _parse_event_stream(body: Iterable[Any]) -> Iterator[dict[str, Any]]:
+        """Yield decoded JSON events from a SageMaker response stream.
+
+        A ``PayloadPart`` is an arbitrary byte chunk that is not aligned to event
+        boundaries: a container such as vLLM or TGI may flush several ``data:`` SSE
+        events in one part, or split a single event across parts. Bytes are buffered;
+        each complete newline-delimited line is decoded as an independent event and any
+        trailing partial line is retained until the next part completes it. Payloads
+        that are not newline-framed (bare JSON) are decoded once the buffer parses whole.
```
Comment (strandly-the-agent) [inline strands-py/src/strands/models/sagemaker.py]: 🟡 **A bare JSON payload that contains newlines is now dropped silently — `main` parsed it, and this docstring says this branch still does.**

The `while "\n" in buffer` loop shreds an indented JSON document into lines and warn-drops each one, so the whole-buffer parse below never sees it. Same mock body, one `PayloadPart` of `json.dumps(event, indent=2)`: `main` → `text='pretty'`, head → `text=''` plus 9 `skipping malformed event` warnings. Identical when split across two parts.

Reachable only via a container that streams indented JSON with no `data:` framing — vLLM/TGI emit compact SSE, DJL/LMI compact jsonlines, and I couldn't name one that does, which is why this isn't a blocker. But it's the same silent loss #4035 is about, so the docs shouldn't promise otherwise.

Suggestion: narrow the claim to match the code (free; `ruff` and all 26 tests stay clean):
```suggestion
        trailing partial line is retained until the next part completes it. A payload that
        is not newline-framed (single-line bare JSON) is decoded once the buffer parses whole.
```
Alternatively, to keep `main`'s behaviour: when a line fails to parse and *isn't* `data:`-prefixed, `break` instead of warn-dropping, leaving the buffer for the whole-buffer parse. I tried it — pretty-printed JSON is restored in both the one-part and split cases, and 26/26 still pass. Cost: a non-`data:` garbage line then poisons the rest of that stream, i.e. the failure mode this PR removes, confined to unframed streams.

**Review by strandly-the-agent — COMMENTED** [review]
**Approve on approach — 3 🟡 should-fixes, nothing blocking.** This fixes real silent data loss (a `PayloadPart` genuinely isn't event-aligned), and the new test really does fail on `main`. Verified on 2571ab676: 26/26 pass, `ruff` clean.

Answering your two questions, `JackYPCOnline`:

**Q1 — "private helper is better right?"** Already done: `_strip_sse_framing` and `_parse_event_stream` are both private `@staticmethod`s, no public surface added. Nothing to change; one optional refinement in Questions below.

**Q2 — "can we resolve this without special-casing it?"** Partly. The SSE half *can* be delegated to `openai._streaming.SSEDecoder` (and `openai>=1.68` is already a dependency of the `sagemaker` extra), which is strictly more spec-correct. But it yields nothing for bare, non-`data:`-framed JSON — the shape two existing tests pin and DJL/LMI containers actually emit. So the SageMaker-specific fallback is irreducible: **the special case can be shrunk, not removed.** Details in the breakdown.

The three should-fixes: one silent behaviour regression vs `main` (multi-line bare JSON, and the docstring promises the opposite), plus two coverage gaps confirmed by mutation — the parser's *other* shipping path and multi-event-per-part beyond N=2.

<details><summary>Evidence ledger — what was actually checked</summary>

- ✅ Branch `pr-4036` @ `2571ab676`, reviewed side-by-side against a `45bcea884` (`main`) worktree; confirmed the old module has no `_parse_event_stream`.
- ✅ `pytest strands-py/tests/strands/models/test_sagemaker.py -q` → **26 passed**. With `main`'s `sagemaker.py` restored under this PR's test file → **1 failed, 25 passed**, the failure being `test_stream_multiple_events_per_payload_part`. It's a true regression test.
- ✅ `ruff format --check` + `ruff check` clean on both changed files.
- ✅ **#4035 repro fixed**: two `data:` events in one `PayloadPart` → `main` `text=''`, head `text='Paris is the capital of France!'` / `end_turn`.
- 🔴 **Regression repro confirmed**: one `PayloadPart` of `json.dumps(event, indent=2)` → `main` `text='pretty'`, head `text=''` + 9 `skipping malformed event` warnings. Same when split across two parts.
- ✅ **Both coverage gaps mutation-verified**: deleting `buffer = ""` (`sagemaker.py:358`) → 26/26 still pass; `while "\n" in buffer` → `if` (`:342`) → 26/26 still pass. Both suggested test fixes pass on head and fail on their mutant.
- ✅ **Pre-existing, identical on both refs** (so out of scope here — see appendix): UTF-8 codepoint split across parts → `UnicodeDecodeError`; `{"choices": []}` → `IndexError`; TGI `{"error": …}` event → `KeyError: 'choices'`; OpenAI-standard top-level `usage` → 0 metadata events emitted.
- ✅ **Reconciled a false lead**: `KeyError: 'PayloadPart'` on a non-`PayloadPart` event isn't reachable — botocore raises `EventStreamError` for the exception members instead of yielding them (`botocore/eventstream.py`, `EventStream._parse_event`). The real gap is that `EventStreamError` isn't in the `except` tuple at `:542-549`; pre-existing.
- ⚠️ Review pipeline note: all specialist passes ran as fresh-context subagents; I could not file the pre-existing bugs as issues (token lacks issue-creation permission), so they're in the appendix for a maintainer to file.

</details>

<details><summary>Q2 in full — how far the special-casing can actually shrink</summary>

Three layers are tangled in `_strip_sse_framing`, and only one is removable:

1. **SSE framing** (`data:` prefix, blank lines, `:` comments, `event:`/`id:`/`retry:`) — genuinely delegatable. `openai._streaming.SSEDecoder().iter_bytes()` handles all of it, including multi-line `data:` fields, correctly. Cost: it's a `_private` symbol of a third-party lib.
2. **The `[DONE]` sentinel** — can't be folded into a generic rule, because after the `data:` strip `[DONE]` still looks like the start of a JSON array.
3. **The bare-JSON fallback** — irreducible, and the reason a generic SSE decoder can't just replace this. `test_stream_with_streaming_enabled` and `test_stream_with_partial_json` both pin unframed JSON with no `data:` prefix; `SSEDecoder` yields nothing for it.

Worth noting the naive version of the rule — "skip any line that isn't a `data:` field" — would *break* those two tests, since bare-JSON lines carry no prefix. If the goal is just to stop the misleading warnings on `event:`/`id:`/`retry:` lines, the narrow version is to return `None` when the payload isn't JSON-looking. See the appendix.

</details>

<details><summary>Questions (1 blocking, 3 non-blocking)</summary>

**❓ Blocking — multi-line bare JSON: fix the behaviour, or narrow the doc?** Either is fine, but one of them needs to happen: right now `sagemaker.py:330` documents behaviour the code doesn't have. The inline comment has both options with measured costs.

**❓ Non-blocking (the Q1 refinement) — module-level `_fn()` instead of `@staticmethod`?** Neither helper touches `self` or `cls`, and `_parse_event_stream` has to spell out `SageMakerAIModel._strip_sse_framing` to reach its sibling — which also quietly defeats any subclass override. Module-level private functions are equally established in this tree, and `bedrock.py:99 _next_stream_event` is the close analogue (a module-level private stream helper). `@classmethod` + `cls.` would work too. Preference, not a rule — happy either way.

**❓ Non-blocking — is the `[DONE]` branch worth keeping?** It's nearly unreachable through `stream()`: the loop `break`s at the first non-None `finish_reason` before the sentinel line is ever parsed, so it only fires when `[DONE]` arrives with no preceding `finish_reason`. Cheap and spec-correct, but it's new untested surface — keep or drop?

**❓ Non-blocking — the `logger.info` → `debug` demotion is a drive-by.** Right direction (that line dumps model content), but it's the very log the issue used as evidence, and `logger.info("choice=<%s>", …)` at `:408` still dumps model content per event at INFO. Keep it here, split it out, or do both?

</details>

<details><summary>Appendix — non-blocking (7 nits + 3 pre-existing bugs worth filing)</summary>

- ⚪ **`assert len(captured_warnings) == 0` (`test_sagemaker.py:493`) isn't load-bearing.** `captured_warnings` is `warnings.catch_warnings` (`tests/conftest.py:129-133`); the parser uses `logger.warning`, which is the correct channel per `strands-py/AGENTS.md`. Verified: a stream with one malformed data line logs `skipping malformed event` and still yields `len(captured_warnings) == 0`. It matches the file's existing convention so it isn't wrong — it just doesn't assert what it looks like it asserts. `caplog.set_level(logging.WARNING, logger="strands.models.sagemaker")` would (pattern at `tests/strands/models/test_anthropic.py:981`).
- ⚪ **The regression test doesn't link #4035**, which root `AGENTS.md` asks for on exactly this kind of test. One line in the docstring: `Regression test for #4035.`
- ⚪ **Spec-legal SSE metadata lines now log as malformed.** `event: message`, `id: 42`, `retry: 1000` each emit `skipping malformed event` (verified: 3 warnings on this branch, 0 on `main`; the payload still parses, so it's log noise, not lost content). Narrow fix: return `None` from `_strip_sse_framing` when the payload isn't JSON-looking — see the Q2 note on why the generic rule doesn't work here.
- ⚪ **`logger.debug("chunk=<%s> | received payload part", buffer)` (`:341`) logs the whole accumulated buffer, not the part it says it received.** Either log the part or rename the key to `buffer=`.
- ⚪ **A stream that dies mid-event still drops its residual silently** (verified: 1 event, 0 warnings) — the PR added a warning for the sibling malformed-line path but not this one, and "no error to correlate" is #4035's own thesis. One line after the `for part in body` loop. Same shape on `main`.
- ⚪ **Test conventions**: `:481` hand-rolls `[chunk async for chunk in …]` where `docs/TESTING.md` asks for the `alist` fixture — already used at `:510` in this same file; and `text`/`message_stop` vs the `tru_`/`exp_` pairing. Only worth touching if you're editing the test anyway.
- ⚪ **`body: Iterable[Any]` (`:322`)** is what lets `part["PayloadPart"]["Bytes"]` past mypy strict; `EventStream[ResponseStreamTypeDef]` is importable and would be the honest annotation.
- 📋 **Pre-existing bugs, verified identical on `main` — not this PR's problem, worth filing** (I couldn't file them from this run):
  1. *UTF-8 codepoint split across `PayloadPart`s* → `UnicodeDecodeError` out of `stream()` (`sagemaker.py:340`; any endpoint emitting unescaped non-ASCII with a part boundary mid-character; ~2-line fix with `codecs.getincrementaldecoder("utf-8")()` now that this PR centralizes the decode).
  2. *Choice-less and error events crash the loop* — `{"choices": [], "usage": {…}}` → `IndexError` and TGI's `data: {"error": …}` → `KeyError: 'choices'` (`sagemaker.py:407`; `openai.py:735` and `llamacpp.py:633` both skip choice-less events); relatedly `EventStreamError` isn't in the `except` tuple at `:542-549`, so mid-stream service failures escape untranslated. This lands on a line the diff touches — folding in a one-line guard here would be a reasonable call if you'd rather not wait.
  3. *Streaming never emits usage metadata for OpenAI-compatible containers* — `if choice.get("usage")` (`:450`) checks inside the choice, but chunks carry `usage` top-level, so vLLM/TGI produce 0 metadata events (the non-streaming path reads top-level `usage` correctly); compounded by the `break` at `:446-448` discarding a trailing usage-only chunk.

</details>

<details><summary>Reading order for human reviewers</summary>

1. **`sagemaker.py:300-320`** — `_strip_sse_framing`, a pure `str -> str | None`. Small, and everything else builds on it.
2. **`sagemaker.py:322-359`** — `_parse_event_stream`, the actual fix. Read the docstring (`:323-330`) against the loop below it; that's where the one behaviour finding lands.
3. **`sagemaker.py:406-461`** — the `stream()` loop. Read this with `git diff -w`: it's a de-indent plus removal of `partial_content` and the `try/except JSONDecodeError`, with no other semantic change (diffed whitespace-blind to confirm).
4. **`tests/strands/models/test_sagemaker.py:456-494`** — the regression test. Fails on `main`, passes here.

Skippable: the `Iterable, Iterator` import at `:6`.

</details>

File: strands-py/tests/strands/models/test_sagemaker.py [hunk 2]
```diff
@@ -452,6 +452,46 @@ async def test_stream_with_partial_json(self, sagemaker_client, model, messages,
...
+        response = [chunk async for chunk in model.stream(messages)]
+
+        text = "".join(
+            event["contentBlockDelta"]["delta"]["text"]
+            for event in response
+            if "contentBlockDelta" in event and "text" in event["contentBlockDelta"]["delta"]
+        )
+        message_stop = next((event for event in response if "messageStop" in event), None)
+
+        assert text == "Paris is the capital of France!"
```
Comment (strandly-the-agent) [inline strands-py/tests/strands/models/test_sagemaker.py]: 🟡 **The bare-JSON multi-part path — the shipping path this rewrite replaced — isn't pinned: delete `buffer = ""` at `sagemaker.py:358` and the parser drops every event after the first, yet all 26 tests still pass.**

`test_stream_with_streaming_enabled` asserts `len(response) >= 5` and `content_delta is not None`, both of which survive that mutant, and `stopReason` stays `end_turn` because `finish_reason` defaults to `""`. Measured on the mutant with that test's own fixture: `text='Paris is the capital of France.'` against `'…France. It is known for the Eiffel Tower.'` on head.

No runtime defect today — the risk is a future refactor silently re-breaking #4035 with a green suite, on the exact shape DJL/LMI containers emit.

Suggestion: assert the reassembled text in `test_stream_with_streaming_enabled` too (its fix lands outside this diff, so no suggestion block):
```python
tru_text = "".join(
    e["contentBlockDelta"]["delta"]["text"]
    for e in response
    if "contentBlockDelta" in e and "text" in e["contentBlockDelta"]["delta"]
)
exp_text = "Paris is the capital of France. It is known for the Eiffel Tower."
assert tru_text == exp_text
```
Verified: passes on head, fails on the mutant. Cost: none — one assertion added to an existing test, which is what `docs/TESTING.md` prefers over a new one.

File: strands-py/tests/strands/models/test_sagemaker.py [hunk 3]
```diff
@@ -452,6 +452,46 @@ async def test_stream_with_partial_json(self, sagemaker_client, model, messages,
...
+        def sse(content: str, finish_reason: str | None = None) -> str:
+            event = {"choices": [{"delta": {"content": content}, "finish_reason": finish_reason}]}
+            return f"data: {json.dumps(event)}\n\n"
+
+        final_event = sse("!", finish_reason="stop")
+        split = len(final_event) // 2
+        mock_response = {
+            "Body": [
+                # Two complete SSE events packed into a single PayloadPart.
+                {"PayloadPart": {"Bytes": (sse("Paris is ") + sse("the capital of France")).encode("utf-8")}},
```
Comment (strandly-the-agent) [inline strands-py/tests/strands/models/test_sagemaker.py]: 🟡 **At N=2, this can't tell the fix apart from a one-event-per-part parser: change `while "\n" in buffer` to `if` at `sagemaker.py:342` and the suite stays green.** The leftover second event gets rescued by the whole-buffer fallback, so the loop that makes multi-event parts work isn't actually pinned.

Same as the other gap — no runtime defect, but "a container flushed several events into one part" *is* #4035, so it's worth pinning at N≥3.

Suggestion: pack three events instead of two — the concatenation is unchanged, so the final assertion still holds:
```suggestion
                {"PayloadPart": {"Bytes": (sse("Paris ") + sse("is the ") + sse("capital of France")).encode("utf-8")}},
```
Verified: 26/26 on head, fails on the `if` mutant, `ruff format`/`check` clean at exactly 120 columns. Cost: none.

**Review by JackYPCOnline — APPROVED** [review]

**Review by mehtarac — APPROVED** [review]
